### DISCUSSION TITLE
Sound effects: use multiple players or change source each time?

### DISCUSSION THREAD

**Discussion by 12people** [bluefireteam/audioplayers#1164] [Q&A]
I have a simple exercise app with 4 different sound effects. One is a countdown, another is the beginning of an exercise, another is the end of the exercise, and the last one represents a successful finish to a workout. There's no instance in which any of these would be played at the same time.

Is it better to use 4 different AudioPlayers, one for each sound? Or should I use one AudioPlayer to change the AssetSource each time?

**Comment by iatndiag** [comment]
I came to the conclusion that I needed 6 * 20 audio players, each 20 are each in their hash table of audio players, six asset directories, each of which has 163 identical one-second WAV files. Only then I do not have a memory leak and in a second I can press the same key at least five times and there will not be a single sound skip. Yes, the temporary directory has grown to 360 MB, but there is no memory leak (140-180 MB is constantly and Not growing, it's great!).

**Reply by iatndiag** [reply]
It turns out that each audio player is tied to its asset source file.

**Reply by iatndiag** [reply]
Approximately like this:
Map<String,AudioPlayer> audioPlayersMap1 = {};  // Map of String:audioplayers, to dispose each one personally by key
Map<String,AudioPlayer> audioPlayersMap2 = {};  // if prev note still sounds, audioplayer is busy, ??? Os: file is busy???
Map<String,AudioPlayer> audioPlayersMap3 = {};  // if prev note still sounds, audioplayer is busy, ??? Os: file is busy???
Map<String,AudioPlayer> audioPlayersMap4 = {};  // if prev note still sounds, audioplayer is busy, ??? Os: file is busy???
Map<String,AudioPlayer> audioPlayersMap5 = {};  // if prev note still sounds, audioplayer is busy, ??? Os: file is busy???
Map<String,AudioPlayer> audioPlayersMap6 = {};  // if prev note still sounds, audioplayer is busy, ??? Os: file is busy???
......
```dart
  break;
  case 4:
    switch (ext) {
      case 1: _xtsn = 'WAV'; _aP = 'wavC/';  break;      case 2: _xtsn = 'm4a'; _aP = 'm4aC/';  break;
      case 3: _xtsn = '';    _aP = '';      break;      case 4: _xtsn = 'mp3'; _aP = 'mp3C/';  break;
      default: {  _xtsn = 'WAV'; _aP = 'wavC/';  } break;
    } //end switch
    _kM = krSnd[_tI][_nI][_sOl].toString();
    _nT = _aP + krSnd[_tI][_nI][_sOl] + '.' + _xtsn; // = we have got a ready path to sound file
    if(audioPlayersMap4.containsKey(_kM)) {} else {
      audioPlayersMap4.addEntries({_kM : AudioPlayer()}.entries);  // adding AudioPlayer for absolutely new note
      audioPlayersMap4 = Map.fromEntries(audioPlayersMap4.entries.toList()..sort((e1, e2) => e1.key.compareTo(e2.key))); // sorting HashMap
    } //end if
    audioPlayersMap4[_kM]?.setVolume(nVol); audioPlayersMap4[_kM]?.play(AssetSource('$_nT'), mode: PlayerMode.lowLatency);
    setState((){countAmap=5;});
    break;
  case 5:
```

**Reply by iatndiag** [reply]
And the solution turned out to be easier! You just need to stop a frequently sounded note. If we press a key quickly again, it stops sounding and starts sounding again. Here is the solution. Just had to add _________________audioPlayersMap[_kM]?.stop();_______________. And just one hash table is enough! All RAM consumption is only 100 MB!!! It's just great! And no memory leak!
```dart
    audioPlayersMap[_kM]?.setVolume(nVol); await audioPlayersMap[_kM]?.stop();
     await audioPlayersMap[_kM]?.play(AssetSource('$_nT'), mode: PlayerMode.lowLatency);

```

**Reply by iatndiag** [reply]
Memory Leak was ONLY in Windows build. In Android build No memory leak At All.

In Android build simple construction works without Memory Leak At All:

```dart
    await player.play(AssetSource('$_nT'), mode: PlayerMode.lowLatency);  // Sounds From Here, Mode for each note: lowLatency ! But Memory Leak Windows!

``` `

**Reply by iatndiag** [reply]
In android, the previous version for Windows will not work. As soon as one and the same note is repeated, there will be no access to the file from the OS side and silence will sound. (The file is in use by another process). But this simple construction works in Android, but causes a memory leak in Windows. Therefore, the previous hash table works well in the Windows assembly.

Solution found!
